fhir-cache/src/main/java/com/ibm/fhir/cache/CachingProxy.java

```diff
@@ -0,0 +1,198 @@
+ */
+public final class CachingProxy {
+    private CachingProxy() { }
+
+    /**
+     * Create a new proxy instance that implements the provided interface and delegates to a caching
+     * invocation handler.
+     *
+     * @param <T>
+     *     the type of the proxy instance
+     * @param interfaceClass
+     *     the interface class
+     * @param target
+     *     the target
+     * @return
+     *     a new proxy instance that implements the provided interface
+     */
+    public static <T> T newInstance(Class<T> interfaceClass, T target) {
+        Objects.requireNonNull(interfaceClass, "interfaceClass");
+        Objects.requireNonNull(target, "target");
+        return interfaceClass.cast(Proxy.newProxyInstance(
+            interfaceClass.getClassLoader(),
+            new Class<?>[] { interfaceClass },
+            new CachingInvocationHandler(target)));
+    }
+
+    /**
+     * Indicates whether the provided class has at least one method with the @Cacheable annotation
+     *
+     * @param targetClass
+     *     the target class
+     * @return
+     *     true if the provided class has at least one method with the @Cacheable annotation, false otherwise
+     */
+    public static boolean hasCacheableMethod(Class<?> targetClass) {
+        Objects.requireNonNull(targetClass, "targetClass");
+        for (Method method : targetClass.getMethods()) {
+            if (isCacheable(method)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * An interface for generating cache keys used by the caching invocation handler to cache method results
+     */
+    public interface KeyGenerator {
+        /**
+         * Default implementation uses the method and argument parameters to generate a cache key
+         */
+        static final KeyGenerator DEFAULT = new KeyGenerator() {
+            @Override
+            public CacheKey generate(Object target, Method method, Object[] args) {
+                return key(method, args);
+            }
+        };
+
+        /**
+         * Generate a cache key for the provided target object, method, and arguments.
+         *
+         * @param target
+         *     the target object
+         * @param method
+         *     the target method
+         * @param args
+         *     the arguments passed to the target method
+         * @return
+         *     a cache key instance
+         */
+        CacheKey generate(Object target, Method method, Object[] args);
+    }
+
+    private static boolean isCacheable(Method method) {
+        return method.isAnnotationPresent(Cacheable.class);
+    }
+
+    private static class CachingInvocationHandler implements InvocationHandler {
+        private static final Object NULL = new Object();
+
+        private final Object target;
+        private final Class<?> targetClass;
+        private final Map<Method, Method> targetMethodCache = new ConcurrentHashMap<>();
```

Comment:
planning to have this use the new CacheSupport stuff so we can have time-based expiry?